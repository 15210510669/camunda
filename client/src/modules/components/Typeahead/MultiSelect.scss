.MultiSelect {
  width: 100%;
  position: relative;

  .UncontrolledMultiValueInput {
    padding-right: 40px;

    .searchClear {
      right: 30px;
    }
  }

  .optionsButton {
    position: absolute;
    right: 0px;
    top: 0;
    height: 100%;
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
    background: none;
    border: none;
    max-height: 32px;
    padding: 0 12px 5px 0;

    .Icon {
      margin-right: 0px;
    }
  }
}
