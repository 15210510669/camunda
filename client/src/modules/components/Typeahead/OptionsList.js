/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {useState, useEffect, useCallback} from 'react';
import {t} from 'translation';
import {Dropdown, LoadingIndicator} from 'components';

export default function OptionsList({
  loading,
  hasMore,
  input,
  onSelect,
  filter,
  children,
  onMouseDown,
  async,
  ...props
}) {
  const [selectedOption, setSelectedOption] = useState(-1);
  const optionList = React.createRef();
  const optionsArr = React.Children.toArray(children);
  let filteredOptions = optionsArr;

  if (!async) {
    filteredOptions = optionsArr.filter(({props: {label, children}}) =>
      (label || children).toLowerCase().includes(filter.toLowerCase())
    );
  }

  const optionsWithProps = filteredOptions.map((option, i) =>
    React.cloneElement(option, {
      className: i === selectedOption ? 'isActive' : '',
      onClick: evt => onSelect(option),
      onMouseDown
    })
  );

  const handleKeyPress = useCallback(
    evt => {
      const {open, onOpen, onClose} = props;
      let nextOption = -1;
      evt = evt || window.event;

      const optionsCount = optionsWithProps.length;

      if (evt.key === 'Enter') {
        evt.stopPropagation();
        const optionToSelect = optionsWithProps[selectedOption];
        if (optionToSelect) {
          onSelect(optionToSelect);
        }
        return;
      }

      if (evt.key === 'Escape') {
        return onClose();
      }

      if (evt.key === 'ArrowDown') {
        if (!open) {
          return onOpen();
        }
        nextOption = (selectedOption + 1) % optionsCount;
      }

      if (evt.key === 'ArrowUp') {
        nextOption = selectedOption - 1 < 0 ? optionsCount - 1 : selectedOption - 1;
      }

      if (optionList.current) {
        const selectedItem = optionList.current.querySelectorAll('.DropdownOption')[nextOption];
        if (selectedItem) {
          selectedItem.scrollIntoView({block: 'nearest', inline: 'nearest'});
        }
      }

      // scroll to end on the last element to show the has more info message
      if (nextOption === optionsCount - 1) {
        optionList.current.scrollTop = optionList.current.scrollHeight;
      }

      setSelectedOption(nextOption);
    },
    [selectedOption, onSelect, optionsWithProps, optionList, props]
  );

  useEffect(() => {
    if (!input) {
      return;
    }
    input.addEventListener('keydown', handleKeyPress);
    return () => {
      input.removeEventListener('keydown', handleKeyPress);
    };
  }, [handleKeyPress, input]);

  if (!props.open) {
    return null;
  }

  const isEmpty = filteredOptions.length === 0 && !loading;

  return (
    <div ref={optionList} className="OptionsList">
      {loading ? <LoadingIndicator /> : optionsWithProps}
      {isEmpty && <Dropdown.Option className="message">{t('common.notFound')}</Dropdown.Option>}
      {hasMore && (
        <Dropdown.Option className="message">
          {t('common.searchForMore', {count: optionsWithProps.length})}
        </Dropdown.Option>
      )}
    </div>
  );
}
