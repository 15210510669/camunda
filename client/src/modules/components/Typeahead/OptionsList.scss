.OptionsList {
  overflow-y: auto;
  overflow-x: hidden;
  border: 1px solid var(--silver-darken-80);
  border-radius: 3px;
  z-index: var(--z-popup);
  position: absolute;
  box-shadow: 0 1px 1px rgba(0, 0, 0, 0.2);
  background-color: var(--silver-lighten-99);
  width: 100%;
  max-height: 350px;

  &:empty {
    display: none;
  }

  .DropdownOption {
    white-space: nowrap;
    overflow-x: hidden;
    text-overflow: ellipsis;
    height: auto;
    &.isActive {
      background-color: rgba(216, 220, 227, 0.5);
      color: var(--grey-darken-13);
      border: none;
      transition: none;
    }

    &:hover {
      background-color: var(--blue-darken-55);
      color: var(--silver-lighten-99);
      .subText {
        color: var(--silver-lighten-99);
      }
    }

    &.message {
      color: var(--grey-lighten-50);
      padding-top: 10px;
      padding-bottom: 10px;

      &:hover,
      &:focus {
        color: var(--grey-lighten-50);
        background-color: var(--silver-lighten-99);
        cursor: auto;
      }
    }
  }
}
