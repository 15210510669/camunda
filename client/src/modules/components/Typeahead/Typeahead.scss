@import 'shared-styles/shared-styles.scss';

.Typeahead {
  width: 100%;
  position: relative;
  .Input {
    width: 100%;
    padding-left: 20px;

    &::placeholder {
      font-style: italic;
      color: #bbbbbb;
    }
  }

  .optionsButton {
    position: absolute;
    right: 0px;
    height: 100%;
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;

    .Icon {
      margin-right: 0px;
    }
  }

  .searchResultsList {
    overflow-y: auto;
    overflow-x: hidden;
    border: 1px solid #bbb;
    border-radius: 3px;
    z-index: 99999;
    position: absolute;
    box-shadow: 0 1px 1px rgba(0, 0, 0, 0.2);
    background-color: white;

    .searchResult {
      &.isActive {
        background-color: rgba(216, 220, 227, 0.5);
        color: black;
        border: none;
        transition: none;
      }

      &:hover {
        background-color: $blue;
        color: #fff;
      }
    }
  }

  ::-webkit-scrollbar {
    -webkit-appearance: none;
    width: 7px;
  }

  ::-webkit-scrollbar-thumb {
    border-radius: 4px;
    background-color: rgba(0, 0, 0, 0.5);
    -webkit-box-shadow: 0 0 1px rgba(255, 255, 255, 0.5);
  }
}
