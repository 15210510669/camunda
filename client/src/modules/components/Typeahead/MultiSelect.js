/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {useRef, useState} from 'react';
import classnames from 'classnames';

import {Button, Icon, Dropdown, MultiValueInput} from 'components';
import {t} from 'translation';

import OptionsList from './OptionsList';
import Typeahead from './Typeahead';

import './MultiSelect.scss';

export default function MultiSelect({
  onSearch = () => {},
  onOpen = () => {},
  onClose = () => {},
  values = [],
  noValuesMessage,
  placeholder,
  className,
  onAdd,
  onRemove,
  async,
  onClear,
  typedOption,
  hasMore,
  loading,
  disabled,
  children,
}) {
  const [open, setOpen] = useState(false);
  const [query, setQuery] = useState('');
  const [optionClicked, setOptionClicked] = useState(false);
  const [resetInput, setResetInput] = useState(false);
  const [insideClick, setInsideClick] = useState(false);

  const input = useRef();

  const isEmpty = !loading && !query && React.Children.count(children) === 0;
  const isInputDisabled = disabled || (isEmpty && !typedOption);

  function showList() {
    if (!open) {
      onOpen(query);
      setOpen(true);
    }
  }

  function hideList() {
    if (!optionClicked) {
      if (query && !insideClick) {
        setResetInput(!resetInput);
        setQuery('');
      }
      onClose();
      setOpen(false);
    } else {
      input.current.focus();
    }
    setInsideClick(false);
  }

  function selectOption({props: {value}}) {
    onAdd(value);
    setOptionClicked(false);
  }

  function getPlaceholderText() {
    if (disabled || isEmpty) {
      return noValuesMessage || t('common.notFound');
    }
    return placeholder;
  }

  function onChange(value) {
    setQuery(value);
    onSearch(value);
  }

  return (
    <div className={classnames('MultiSelect', className)} onMouseDown={() => setInsideClick(true)}>
      <MultiValueInput
        className="typeaheadInput"
        key={resetInput}
        onChange={onChange}
        values={values}
        onAdd={(...args) => {
          if (!optionClicked) {
            onAdd(...args);
          }
        }}
        onRemove={onRemove}
        ref={input}
        onBlur={hideList}
        onClear={onClear}
        onFocus={showList}
        disabled={isInputDisabled}
        placeholder={getPlaceholderText(isEmpty)}
        disableAddByKeyboard
      />
      <Button
        tabIndex="-1"
        className="optionsButton"
        onClick={() => input.current.focus()}
        disabled={isInputDisabled}
      >
        <Icon type="down" className="downIcon" />
      </Button>
      <OptionsList
        async={async}
        open={open}
        onClose={hideList}
        onOpen={open}
        filter={query}
        onSelect={selectOption}
        input={input.current}
        onMouseDown={() => setOptionClicked(true)}
        loading={loading}
        hasMore={hasMore}
        typedOption={typedOption}
      >
        {children}
      </OptionsList>
    </div>
  );
}

MultiSelect.Option = Dropdown.Option;

MultiSelect.Highlight = Typeahead.Highlight;
