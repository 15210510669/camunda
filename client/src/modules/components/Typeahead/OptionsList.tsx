/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {
  useState,
  useEffect,
  useCallback,
  ReactNode,
  ReactElement,
  MouseEvent,
  isValidElement,
  cloneElement,
  createRef,
  Children,
} from 'react';

import {t} from 'translation';
import {Dropdown, LoadingIndicator} from 'components';

import {highlightText} from './service';

import './OptionsList.scss';

export interface OptionProps {
  label: string;
  value: string;
  children: ReactNode;
  disabled?: boolean;
  className: string;
  onClick: () => void;
  onMouseDown?: (evt: MouseEvent<HTMLDivElement, MouseEvent>) => void;
}

interface OptionsListProps {
  loading?: boolean;
  hasMore?: boolean;
  input: HTMLInputElement | null;
  onSelect: (option: ReactElement<OptionProps>) => void;
  filter?: string;
  children: ReactNode;
  onMouseDown?: (evt: MouseEvent<HTMLDivElement, MouseEvent>) => void;
  async?: boolean;
  typedOption?: boolean;
  open: boolean;
  onOpen: () => void;
  onClose: () => void;
}

export default function OptionsList({
  loading,
  hasMore,
  input,
  onSelect,
  filter,
  children,
  onMouseDown,
  async,
  typedOption,
  ...props
}: OptionsListProps): JSX.Element | null {
  const [selectedOption, setSelectedOption] = useState<number>(-1);
  const optionList = createRef<HTMLDivElement>();
  const optionsArr = Children.toArray(children).filter(isValidElement<OptionProps>);
  let filteredOptions: ReactElement<OptionProps>[] = optionsArr;

  if (!async && filter) {
    filteredOptions = optionsArr.filter(({props: {label, children}}) => {
      if (filter && typedOption && (label || children) === filter) {
        return false; // remove options that exactly match the typed option
      }
      return (label || children)?.toString().toLowerCase().includes(filter.toLowerCase());
    });
  }

  if (filter && typedOption) {
    filteredOptions.unshift(
      <Dropdown.Option key={filter} value={filter} label={filter}>
        {filter}
      </Dropdown.Option>
    );
  }

  const optionsWithProps = filteredOptions.map((option, i) =>
    cloneElement(option, {
      className: i === selectedOption ? 'isActive' : '',
      onClick: () => onSelect(option),
      onMouseDown: (evt: MouseEvent<HTMLDivElement, MouseEvent>) =>
        !option.props.disabled && onMouseDown?.(evt),
      children: highlightText(option.props.children, filter),
    })
  );

  const handleKeyPress = useCallback(
    (evt: KeyboardEvent) => {
      const {open, onOpen, onClose} = props;
      let nextOption = -1;
      evt = evt || window.event;

      const optionsCount = filteredOptions.length;

      if (evt.key === 'Enter') {
        evt.stopPropagation();
        const optionToSelect = filteredOptions[selectedOption];
        if (optionToSelect && !optionToSelect.props.disabled) {
          onSelect(optionToSelect);
        }
        return;
      }

      if (evt.key === 'Escape') {
        return onClose();
      }

      if (evt.key === 'ArrowDown') {
        if (!open) {
          return onOpen();
        }
        nextOption = (selectedOption + 1) % optionsCount;
      }

      if (evt.key === 'ArrowUp') {
        nextOption = selectedOption - 1 < 0 ? optionsCount - 1 : selectedOption - 1;
      }

      if (optionList.current) {
        const selectedItem = optionList.current.querySelectorAll('.DropdownOption')[nextOption];
        if (selectedItem) {
          selectedItem.scrollIntoView({block: 'nearest', inline: 'nearest'});
        }

        // scroll to end on the last element to show the has more info message
        if (nextOption === optionsCount - 1) {
          optionList.current.scrollTop = optionList.current.scrollHeight;
        }
      }

      setSelectedOption(nextOption);
    },
    [selectedOption, onSelect, filteredOptions, optionList, props]
  );

  useEffect(() => {
    if (!input) {
      return;
    }
    input.addEventListener('keydown', handleKeyPress);
    return () => {
      input.removeEventListener('keydown', handleKeyPress);
    };
  }, [handleKeyPress, input]);

  if (!props.open) {
    return null;
  }

  const isEmpty = filteredOptions.length === 0 && !loading && !typedOption;

  return (
    <div ref={optionList} className="OptionsList">
      {loading ? <LoadingIndicator /> : optionsWithProps}
      {isEmpty && <Dropdown.Option className="message">{t('common.notFound')}</Dropdown.Option>}
      {hasMore && (
        <Dropdown.Option className="message">
          {t('common.searchForMore', {count: optionsWithProps.length})}
        </Dropdown.Option>
      )}
    </div>
  );
}
