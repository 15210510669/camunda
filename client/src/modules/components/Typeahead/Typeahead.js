/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import classnames from 'classnames';
import {Input, Dropdown} from 'components';
import {formatters} from 'services';

import './Typeahead.scss';

const valuesShownInBox = 10;
const valueHeight = 30;

export default class Typeahead extends React.Component {
  state = {
    query: '',
    optionsVisible: false,
    selectedValueIdx: 0,
    firstShownOptionIdx: 0,
    lastCommittedValue: ''
  };

  componentDidMount() {
    document.body.addEventListener('click', this.close);
  }

  componentDidUpdate(prevProps) {
    if (this.props.initialValue && this.props.initialValue !== prevProps.initialValue) {
      const value = this.getFormatter()(this.props.initialValue);
      this.setState({
        query: value,
        lastCommittedValue: value
      });
    }
    if (this.props.disabled !== prevProps.disabled) {
      this.setState({
        query: this.props.disabled ? this.props.placeholder : '',
        lastCommittedValue: ''
      });
    }
  }

  componentWillUnmount() {
    document.body.removeEventListener('click', this.close);
  }

  returnFocusToInput = evt => {
    if (evt.target === this.optionsList) {
      this.input.focus();
    }
  };

  showOptions = evt => {
    if (evt && evt.type === 'click') {
      this.input.select();
    }

    if (!this.state.optionsVisible) {
      this.setState({
        optionsVisible: true
      });
    }
  };

  close = evt => {
    if (!evt || !this.container.contains(evt.target)) {
      this.setState({
        optionsVisible: false
      });
    }
  };

  updateValues = () => {
    this.setState({
      selectedValueIdx: 0,
      firstShownOptionIdx: 0
    });
  };

  updateQuery = async evt => {
    this.setState({query: evt.target.value});
    this.showOptions();
    this.updateValues();
  };

  getFormatter = () => this.props.formatter || (v => v);

  selectValue = value => () => {
    const formatter = this.getFormatter();
    const formattedValue = formatter(value);
    this.setState({
      query: formattedValue,
      lastCommittedValue: formattedValue,
      selectedValueIdx: 0,
      firstShownOptionIdx: 0,
      optionsVisible: false
    });

    if (this.props.onSelect) {
      this.props.onSelect(value);
    }
    this.close();
  };

  resetToLastCommitted = () => {
    this.setState({query: this.state.lastCommittedValue});
  };

  handleKeyPress = evt => {
    evt = evt || window.event;

    if (evt.key === 'Tab') {
      this.close();
      return;
    }

    const {selectedValueIdx, optionsVisible} = this.state;
    const values = this.getFilteredValues();
    if (evt.key === 'Enter') {
      evt.preventDefault();
      if (optionsVisible && values[selectedValueIdx]) {
        this.selectValue(values[selectedValueIdx])();
        evt.stopPropagation();
      }
      return;
    }

    if (evt.key === 'Escape') {
      if (this.state.optionsVisible) {
        evt.stopPropagation();
      }
      this.close();
    } else {
      let selectedValueIdx = this.state.selectedValueIdx;

      if (evt.key === 'ArrowDown') {
        evt.preventDefault();
        if (!this.state.optionsVisible) {
          this.showOptions();
        } else {
          selectedValueIdx = (selectedValueIdx + 1) % values.length;
        }
      }

      if (evt.key === 'ArrowUp') {
        evt.preventDefault();
        selectedValueIdx = selectedValueIdx - 1 < 0 ? values.length - 1 : selectedValueIdx - 1;
      }

      const firstShownOptionIdx = this.scrollIntoView(selectedValueIdx);

      this.setState({
        selectedValueIdx,
        firstShownOptionIdx
      });
    }
  };

  scrollIntoView = selectedValueIdx => {
    let {firstShownOptionIdx} = this.state;
    const values = this.getFilteredValues();

    if (this.optionsList) {
      if (selectedValueIdx === 0) {
        this.optionsList.scrollTop = 0;
        firstShownOptionIdx = 0;
      }

      if (selectedValueIdx === values.length - 1) {
        this.optionsList.scrollTop = this.optionsList.scrollHeight;
        firstShownOptionIdx = values.length - valuesShownInBox;
      }

      if (selectedValueIdx >= firstShownOptionIdx + valuesShownInBox) {
        this.optionsList.scrollTop = (firstShownOptionIdx + 1) * valueHeight;
        firstShownOptionIdx++;
      }

      if (selectedValueIdx < firstShownOptionIdx) {
        firstShownOptionIdx--;
        this.optionsList.scrollTop = selectedValueIdx * valueHeight;
      }
    }

    return firstShownOptionIdx;
  };

  optionsListRef = optionsList => {
    this.optionsList = optionsList;
    if (optionsList) {
      this.optionsList.highestWidth = this.optionsList.offsetWidth;
    }
  };

  containerRef = container => {
    this.container = container;
  };

  inputRef = input => {
    this.input = input;
  };

  getFilteredValues = () =>
    this.props.values.filter(value =>
      this.getFormatter()(value)
        .toLowerCase()
        .includes(this.state.query.toLowerCase())
    );

  render() {
    const formatter = this.getFormatter();

    const {query, selectedValueIdx, optionsVisible} = this.state;
    const values = this.getFilteredValues();

    const searchResultContainerStyle = {
      maxHeight: valueHeight * valuesShownInBox + 'px',
      maxWidth: this.optionsList && this.optionsList.highestWidth + 'px'
    };

    const valueStyle = {
      height: valueHeight + 'px',
      minWidth: this.input && this.input.clientWidth + 'px'
    };

    const hasNoValues = this.props.values.length === 0;

    return (
      <div ref={this.containerRef} className={classnames('Typeahead', this.props.className)}>
        <Input
          className={classnames({isInvalid: this.props.isInvalid})}
          value={hasNoValues ? 'No results found' : query}
          onChange={this.updateQuery}
          onClick={this.showOptions}
          onFocus={this.showOptions}
          onKeyDown={this.handleKeyPress}
          onBlur={this.resetToLastCommitted}
          placeholder={this.props.placeholder}
          ref={this.inputRef}
          disabled={this.props.disabled || hasNoValues}
        />
        {optionsVisible && values.length > 0 && (
          <div
            style={searchResultContainerStyle}
            className="searchResultsList"
            ref={this.optionsListRef}
            onMouseUp={this.returnFocusToInput}
          >
            {values.map((value, index) => {
              return (
                <Dropdown.Option
                  className={classnames('searchResult', {
                    isActive: index === selectedValueIdx
                  })}
                  style={valueStyle}
                  onMouseDown={evt => evt.preventDefault()}
                  onClick={this.selectValue(value)}
                  key={value && value.id ? value.id : formatter(value)}
                >
                  {formatters.getHighlightedText(formatter(value), this.state.query)}
                </Dropdown.Option>
              );
            })}
          </div>
        )}
      </div>
    );
  }
}
