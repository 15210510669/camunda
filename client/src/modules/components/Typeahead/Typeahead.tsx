/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {
  ReactElement,
  createRef,
  ChangeEvent,
  Component,
  Children,
  isValidElement,
  ReactNode,
} from 'react';

import {Input, Button, Icon, Dropdown} from 'components';
import {t} from 'translation';
import classnames from 'classnames';

import OptionsList from './OptionsList';

import './Typeahead.scss';

interface TypeaheadProps {
  children: ReactNode;
  initialValue?: string;
  value?: string;
  onSearch: (query: string) => void;
  onChange: (value: string) => void;
  onOpen: () => void;
  onClose: () => void;
  placeholder?: string | JSX.Element[];
  disabled?: boolean;
  loading?: boolean;
  hasMore?: boolean;
  async?: boolean;
  typedOption?: boolean;
  noValuesMessage?: string | JSX.Element[];
  className?: string;
}

interface TypeaheadState {
  query: string;
  open: boolean;
  selected: string | undefined;
}

const defaultState: TypeaheadState = {
  query: '',
  open: false,
  selected: '',
};

export default class Typeahead extends Component<TypeaheadProps, TypeaheadState> {
  static defaultProps = {
    onSearch: () => {},
    onChange: () => {},
    onOpen: () => {},
    onClose: () => {},
  };

  state = defaultState;

  optionClicked = false;

  input = createRef<HTMLInputElement>();

  onTextChange = (evt: ChangeEvent<HTMLInputElement>) => {
    this.setState({query: evt.target.value, open: true});
    this.props.onSearch(evt.target.value);
  };

  componentDidMount() {
    const {initialValue, value} = this.props;

    this.findAndSelect(initialValue);
    this.findAndSelect(value);
  }

  componentDidUpdate(prevProps: TypeaheadProps) {
    if (prevProps.value !== this.props.value) {
      if (typeof this.props.value === 'undefined') {
        this.setState(defaultState);
        this.props.onSearch('');
      } else {
        this.findAndSelect(this.props.value);
      }
    }
    if (!prevProps.initialValue && this.props.initialValue) {
      this.findAndSelect(this.props.initialValue);
    }
  }

  findAndSelect = (value?: string) => {
    const {children, typedOption} = this.props;
    const foundOption = Children.toArray(children).find(
      (option) => (option as ReactElement).props.value === value
    );

    if (typedOption) {
      this.setState({
        selected: value,
        query: value || '',
      });
    } else if (isValidElement(foundOption)) {
      const {label, children} = foundOption.props;
      const selected = label || children;
      this.setState({
        selected,
        query: selected,
      });
    }
  };

  selectOption = ({props: {label, children, value}}: ReactElement) => {
    const selected = label || children;
    this.setState({
      selected,
      query: selected,
      open: false,
    });
    this.props.onChange(value);
    this.optionClicked = false;
  };

  getPlaceholderText = (isEmpty: boolean) => {
    const {placeholder, disabled, noValuesMessage} = this.props;
    const {selected} = this.state;
    if (selected) {
      return selected;
    }
    if (disabled || isEmpty) {
      return noValuesMessage || t('common.notFound');
    }
    return placeholder;
  };

  open = () => {
    this.props.onOpen();
    this.setState({open: true, query: ''});
  };

  close = () => {
    this.props.onClose();
    this.setState(({selected}) => ({open: false, query: selected || ''}));
  };

  render() {
    const {children, disabled, loading, hasMore, async, typedOption, className} = this.props;
    const {query, open, selected} = this.state;
    const isEmpty = !loading && !query && !typedOption && Children.count(children) === 0;
    const isInputDisabled = isEmpty || disabled;

    return (
      <div className={classnames(className, 'Typeahead')}>
        <Input
          type="text"
          className={classnames('typeaheadInput', {selectionVisible: open && selected})}
          value={query}
          onFocus={this.open}
          onBlur={(evt) => {
            if (!this.optionClicked) {
              this.close();
            }
          }}
          onChange={this.onTextChange}
          ref={this.input}
          placeholder={this.getPlaceholderText(isEmpty)}
          disabled={isInputDisabled}
        />
        <Button
          tabIndex={-1}
          className="optionsButton"
          onClick={() => this.input.current?.focus()}
          disabled={isInputDisabled}
        >
          <Icon type="down" className="downIcon" />
        </Button>
        <OptionsList
          async={async}
          open={open}
          onClose={this.close}
          onOpen={this.open}
          filter={query}
          onSelect={this.selectOption}
          input={this.input.current}
          onMouseDown={() => (this.optionClicked = true)}
          loading={loading}
          hasMore={hasMore}
          typedOption={typedOption}
        >
          {children}
        </OptionsList>
      </div>
    );
  }

  static Option = Dropdown.Option;

  static Highlight = function Highlight(props: {children: string; matchFromStart?: boolean}) {
    return <>{props.children}</>;
  };
}
