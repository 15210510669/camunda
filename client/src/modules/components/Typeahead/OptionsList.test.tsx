/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {shallow} from 'enzyme';
import {runLastEffect} from '__mocks__/react';
import {Dropdown} from 'components';

import OptionsList from './OptionsList';
import Typeahead from './Typeahead';

const props = {
  open: true,
  input: {
    addEventListener: jest.fn(),
  } as jest.MockedObject<HTMLInputElement>,
  filter: '',
  onSelect: jest.fn(),
  children: <div />,
  onOpen: jest.fn(),
  onRemove: jest.fn(),
  onClose: () => {},
};

it('should render an empty OptionsList', () => {
  const node = shallow(<OptionsList {...props} />);

  expect(node).toMatchSnapshot();
});

it('should attach keyPressEvent to input', () => {
  shallow(<OptionsList {...props} />);

  runLastEffect();

  expect(props.input.addEventListener).toHaveBeenCalled();
});

it('should filter option list options', () => {
  const node = shallow(
    <OptionsList {...props} filter="one">
      <Typeahead.Option id="test_option" value="1">
        One
      </Typeahead.Option>
      <Typeahead.Option id="second_option" value="2">
        Two
      </Typeahead.Option>
    </OptionsList>
  );

  expect(node.find('#second_option')).not.toExist();
});

it('should invoke onSelect when selecting an option', () => {
  const node = shallow(
    <OptionsList {...props}>
      <Typeahead.Option id="test_option" value="1">
        One
      </Typeahead.Option>
      <Typeahead.Option id="second_option" value="2">
        Two
      </Typeahead.Option>
    </OptionsList>
  );

  node.find('#second_option').simulate('click');

  expect(props.onSelect.mock.calls[0][0].props.value).toBe('2');
});

it('should render has more info message if specified', () => {
  const node = shallow(<OptionsList {...props} hasMore={true} />);

  expect(node).toMatchSnapshot();
});

it('should render an option with the search as input if optionList has typedOption prop and empty', () => {
  const node = shallow(<OptionsList {...props} filter="test" typedOption />);

  expect(node).toMatchSnapshot();
});

it('should not display an option if it is identical to the typed option', () => {
  const node = shallow(
    <OptionsList {...props} filter="test" typedOption>
      <Typeahead.Option id="test_option" value="1">
        Test 1
      </Typeahead.Option>
      <Typeahead.Option id="second_option" value="2">
        test
      </Typeahead.Option>
    </OptionsList>
  );

  expect(node.find(Dropdown.Option)).toHaveLength(2);
});

it('should not crash when not providing a mouseDown handler', () => {
  const node = shallow(
    <OptionsList {...props}>
      <Typeahead.Option id="test_option" value="1">
        Test 1
      </Typeahead.Option>
    </OptionsList>
  );

  node.find(Dropdown.Option).simulate('mousedown');
});

it('should not crash when providing null children', () => {
  const node = shallow(
    <OptionsList {...props}>
      <Typeahead.Option id="test_option" value="1">
        {null}
      </Typeahead.Option>
    </OptionsList>
  );

  node.find(Dropdown.Option).simulate('mousedown');
});
