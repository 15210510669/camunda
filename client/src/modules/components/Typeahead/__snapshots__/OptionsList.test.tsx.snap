// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an empty OptionsList 1`] = `
<div
  className="OptionsList"
>
  <div
    className=""
    key=".0"
    onClick={[Function]}
    onMouseDown={[Function]}
  />
</div>
`;

exports[`should render an option with the search as input if optionList has typedOption prop and empty 1`] = `
<div
  className="OptionsList"
>
  <ForwardRef(DropdownOption)
    className=""
    key="test"
    label="test"
    onClick={[Function]}
    onMouseDown={[Function]}
    value="test"
  >
    <span
      key="0"
    />
    <span
      className="textBold"
      key="1"
    >
      test
    </span>
    <span
      key="2"
    />
  </ForwardRef(DropdownOption)>
</div>
`;

exports[`should render has more info message if specified 1`] = `
<div
  className="OptionsList"
>
  <div
    className=""
    key=".0"
    onClick={[Function]}
    onMouseDown={[Function]}
  />
  <ForwardRef(DropdownOption)
    className="message"
  >
    Displaying 1 results only. Search to refine.
  </ForwardRef(DropdownOption)>
</div>
`;
